x-pack/auditbeat/module/system/socket/state.go

```diff
@@ -647,23 +633,31 @@ func (s *state) ThreadLeave(tid uint32) (ev event, found bool) {
 	if ev, found = s.threads[tid]; found {
 		delete(s.threads, tid)
 	}
-	return
+	return ev, found
 }
 
-func (s *state) onSockTerminated(sock *socket) {
-	for _, flow := range sock.flows {
-		s.onFlowTerminated(flow)
+func (s *state) onSockTerminated(sock *socket) (toReport linkedList) {
+	for _, f := range sock.flows {
+		flows := s.onFlowTerminated(f)
+		toReport.append(&flows)
 	}
+	sock.flows = nil
 	delete(s.socks, sock.sock)
 	if sock.closing {
 		s.closing.remove(sock)
 	} else {
 		s.moveToClosing(sock)
 	}
+	return toReport
 }
 
 // CreateSocket allocates a new sock in the system
 func (s *state) CreateSocket(ref flow) error {
+	var toReport linkedList
+	// Send flows to the output as a deferred function to avoid
+	// holding on s mutex when there's backpressure from the output.
+	defer s.reportFlows(&toReport)
```

Comment:
I didn't want to log in the main thread as this will report a flow every time a socket is closed.

The logging in ExpireFlows is there to have a way to tell when backpressure is happening without causing too much impact.